### PR TITLE
Make sure that SettingFile arg is parsed before we load the settings

### PR DESCRIPTION
**Pull request by TravisEz13**
## PR Summary

Make sure that SettingFile arg is parsed before we load the settings
  - Add ability to parse some setting before console host is created
  - Trigger the Parse before the console host is created
  - disable snapin loading from loading setting before this point
  - re-enable mac logging tests.

Also, fix the following Travis-ci issues:

- make sure the Linux and macOS caches are separate to reduce the size
- update the macOS image to the closest image as used by the official build system (equivalent to VSTS hosted mac builds)

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 1]
```diff
@@ -392,6 +393,218 @@ internal string WorkingDirectory
...
+
+        private static string GetConfigurationNameFromGroupPolicy()
+        {
+            // Current user policy takes precedence.
+            var consoleSessionSetting = Utils.GetPolicySetting<ConsoleSessionConfiguration>(Utils.CurrentUserThenSystemWideConfig);
+            if (consoleSessionSetting != null)
+            {
+                if (consoleSessionSetting.EnableConsoleSessionConfiguration == true)
+                {
+                    if (!string.IsNullOrEmpty(consoleSessionSetting.ConsoleSessionConfigurationName))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: We could combine three `if` in one or use `?` operator in just single `return`.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 2]
```diff
@@ -392,6 +393,218 @@ internal string WorkingDirectory
...
+                {
+                    break;
+                }
+
+                string switchKey = switchKeyResults.SwitchKey;
+
+                if (MatchSwitch(switchKey, "settingsfile", "settings"))
+                {
+                    // parse setting file arg and don't write error as there is no host yet.
+                    if (!ParseSettingFileHelper(args, ++i, null))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: Maybe use `TryParseSettingFileHelper`?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 3]
```diff
@@ -392,6 +393,218 @@ internal string WorkingDirectory
...
+            Dbg.Assert(!String.IsNullOrEmpty(match), "need a value");
+            Dbg.Assert(match.Trim().ToLowerInvariant() == match, "match should be normalized to lowercase w/ no outside whitespace");
+            Dbg.Assert(smallestUnambiguousMatch.Trim().ToLowerInvariant() == smallestUnambiguousMatch, "match should be normalized to lowercase w/ no outside whitespace");
+            Dbg.Assert(match.Contains(smallestUnambiguousMatch), "sUM should be a substring of match");
+
+            if (match.Trim().ToLowerInvariant().IndexOf(switchKey, StringComparison.Ordinal) == 0)
+            {
+                if (switchKey.Length >= smallestUnambiguousMatch.Length)
+                {
+                    return true;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: We could exclude if-s and use single `return`.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs [hunk 4]
```diff
@@ -148,6 +148,12 @@ internal sealed partial class ConsoleHost
 
             try
             {
+                // We need to read the settings file before we create the console host
+                string[] tempArgs = new string[args.GetLength(0)];
+                args.CopyTo(tempArgs, 0);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Could you please clarify why we need clone `args`?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 5]
```diff
@@ -200,6 +200,7 @@ internal CommandLineParameterParser(PSHostUserInterface hostUI, string bannerTex
             _helpText = helpText;
         }
 
+#region Internal properties
         internal bool AbortStartup
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: Should `#region` be in position with the line?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 4 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: The existing code cloned the args.  I don't want to break something by not cloning.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 1 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: `?` only checks for null and could have regressions.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 2 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: updated

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 3 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: sorry, I don't follow what you are saying.

Well, looking at the code, I made a guess.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 5 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: updated

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 4 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Reviewing the code.  I don't think it has to be.  I especially don't think it has to be twice.  compromise, I'll remove the second clone, you file an issue to remove the first one.

**Comment by TravisEz13** [comment]
@iSazonov  I believe I have addressed all your comments one way or another.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 4 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Done #7457.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 6]
```diff
@@ -393,6 +394,203 @@ internal string WorkingDirectory
...
+            return (SwitchKey: switchKey, ShouldBreak: false);
+        }
+
+        private static string NormalizeFilePath(string path)
+        {
+            // Normalize slashes
+            path = path.Replace(StringLiterals.AlternatePathSeparator,
+                                StringLiterals.DefaultPathSeparator);
+
+            return Path.GetFullPath(path);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: The method is used once and then we call File.Exist() where Path.GetFullPath() is called again.
https://source.dot.net/#System.IO.FileSystem/System/IO/File.cs,119
We could simplify the TryParseSettingFileHelper() method - remove try-catch there (line 429), move the line 574 to TryParseSettingFileHelper()  before File.Exist()

File: test/powershell/Host/Logging.Tests.ps1 [hunk 7]
```diff
@@ -201,34 +201,33 @@ Describe 'Basic os_log tests on MacOS' -Tag @('CI','RequireSudoOnUnix') {
...
 
-    # It 'Verifies logging level filtering works' -Skip:(!$IsSupportedEnvironment) {
-    It 'Verifies logging level filtering works' -Pending {
+    It 'Verifies logging level filtering works' -Skip:(!$IsSupportedEnvironment) {
         $configFile = WriteLogSettings -LogId $logId -LogLevel Warning
-        & $powershell -NoProfile -SettingsFile $configFile -Command '$env:PSModulePath | out-null'
+        $testPid = & $powershell -NoLogo -NoProfile -SettingsFile $configFile -Command '$PID'
 
-        Export-PSOsLog -After $after -Verbose | Set-Content -Path $contentFile
+        Export-PSOsLog -after $after -LogPid $testPid -Verbose | Set-Content -Path $contentFile
```
Comment (iSazonov) [inline test/powershell/Host/Logging.Tests.ps1]: Please revert `-after` to `-After`.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 8]
```diff
@@ -393,6 +394,203 @@ internal string WorkingDirectory
...
+
+                return false;
+            }
+
+            string configFile = null;
+            try
+            {
+                configFile = NormalizeFilePath(args[argIndex]);
+            }
+            catch (Exception ex)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: I believe we could remove the `try-catch`. See my comment for NormalizeFilePath() method.

**Review by iSazonov — COMMENTED** [review]
Leave a comment

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 3 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: I meant 

```
return (match.Trim().ToLowerInvariant().IndexOf(switchKey, StringComparison.Ordinal) == 0) && (switchKey.Length >= smallestUnambiguousMatch.Length)

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 9]
```diff
@@ -393,6 +394,203 @@ internal string WorkingDirectory
...
+            EarlyParseHelper(args);
+        }
+
+        private static string GetConfigurationNameFromGroupPolicy()
+        {
+            // Current user policy takes precedence.
+            var consoleSessionSetting = Utils.GetPolicySetting<ConsoleSessionConfiguration>(Utils.CurrentUserThenSystemWideConfig);
+
+            return (consoleSessionSetting?.EnableConsoleSessionConfiguration == true && !string.IsNullOrEmpty(consoleSessionSetting?.ConsoleSessionConfigurationName)) ? 
+                    consoleSessionSetting.ConsoleSessionConfigurationName : string.Empty;
```
[the remark this answers is not in the document]
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: this is an existing variable name.  Please file an issues

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 10]
```diff
@@ -393,6 +394,203 @@ internal string WorkingDirectory
...
+        /// <param name="args">
+        /// The command line parameters to be processed.
+        /// </param>
+        private static void EarlyParseHelper(string[] args)
+        {
+            Dbg.Assert(args != null, "Argument 'args' to ParseHelper should never be null");
+            bool noexitSeen = false;
+            for (int i = 0; i < args.Length; ++i)
+            {
+                var switchKeyResults = GetSwitchKey(args, ref i, null, ref noexitSeen);
```
[the remark this answers is not in the document]
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: resolved

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 11]
```diff
@@ -393,6 +394,203 @@ internal string WorkingDirectory
...
+                if (switchKeyResults.ShouldBreak)
+                {
+                    break;
+                }
+
+                string switchKey = switchKeyResults.SwitchKey;
+
+                if (MatchSwitch(switchKey, "settingsfile", "settings"))
```
[the remark this answers is not in the document]
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: resolved

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 6 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: existing code, please file an issue, the method was just changed from an instance to static

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs [hunk 12]
```diff
@@ -148,6 +148,12 @@ internal sealed partial class ConsoleHost
...
+
+                CommandLineParameterParser.EarlyParse(tempArgs);
```
Comment (dantraMSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: I've created issue https://github.com/PowerShell/PowerShell/issues/7460 to move this call to UnmanagedPSEntry.Start prior to PSEtwLog.LogConsoleStartup().

File: test/powershell/Host/Logging.Tests.ps1 [hunk 13]
```diff
@@ -201,34 +201,33 @@ Describe 'Basic os_log tests on MacOS' -Tag @('CI','RequireSudoOnUnix') {
...
-    It 'Verifies basic logging with no customizations' -Pending {
+    It 'Verifies basic logging with no customizations' -Skip:(!$IsSupportedEnvironment) {
         $configFile = WriteLogSettings -LogId $logId
-        & $powershell -NoProfile -SettingsFile $configFile -Command '$env:PSModulePath | out-null'
+        $testPid = & $powershell -NoProfile -SettingsFile $configFile -Command '$PID'
 
-        Export-PSOsLog -After $after -Verbose | Set-Content -Path $contentFile
-        $items = Get-PSOsLog -Path $contentFile -Id $logId -After $after -TotalCount 3 -Verbose
+        # Made tests more reliable
+        Start-Sleep -Milliseconds 500
```
Comment (dantraMSFT) [inline test/powershell/Host/Logging.Tests.ps1]: I've created issue https://github.com/PowerShell/PowerShell/issues/7462 to address the Start-Sleep workaround as well as the possible false-positive in filtering test below.

**Review by dantraMSFT — APPROVED** [review]
None of my comments are blocking but recommend you reconsider adding release logic for the new Dbg.Assert calls.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 14]
```diff
@@ -463,57 +661,20 @@ internal void Parse(string[] args)
...
             bool noexitSeen = false;
 
             for (int i = 0; i < args.Length; ++i)
             {
-                // Invariant culture used because command-line parameters are not localized.
-
-                string switchKey = args[i].Trim().ToLowerInvariant();
-                if (String.IsNullOrEmpty(switchKey))
+                (string switchKey, bool shouldBreak) switchKeyResults = GetSwitchKey(args, ref i, this, ref noexitSeen);
+                if (switchKeyResults.ShouldBreak)
```
Comment (dantraMSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: I'm pretty sure tuple field names are case sensitive.

File: .travis.yml [hunk 15]
```diff
@@ -8,8 +8,10 @@ matrix:
     - os: linux
       dist: trusty
       sudo: required
+      env: TRAVIS_CI_CACHE_NAME=linux
     - os: osx
-      osx_image: xcode8.1
+      osx_image: xcode9.4
```
[the remark this answers is not in the document]
Comment (TravisEz13) [reply .travis.yml]: I found no differences in the log format, but yes correct

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 12 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Thanks

File: test/powershell/Host/Logging.Tests.ps1 (hunk 13 above)
Comment (TravisEz13) [reply test/powershell/Host/Logging.Tests.ps1]: Thanks

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 14 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: updated

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 16]
```diff
@@ -393,6 +394,203 @@ internal string WorkingDirectory
...
+        /// For example -foo, /foo, or --foo would return 'foo'.
+        /// </summary>
+        /// <param name="args">
+        /// The command line parameters to be processed.
+        /// </param>
+        /// <param name="argIndex">
+        /// The index in args to the argument to process.
+        /// </param>
+        /// <param name="parser">
+        /// Used to allow the helper to write errors to the console.  If not supplied, Files will not be parsed.
```
[the remark this answers is not in the document]
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: Updated comment (I have not pushed at this point.)

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 17]
```diff
@@ -393,6 +394,203 @@ internal string WorkingDirectory
...
+        /// <summary>
+        /// Processes the command line parameters to ConsoleHost which must be parsed before the Host is created.
+        /// Success to indicate that the program should continue running.
+        /// </summary>
+        /// <param name="args">
+        /// The command line parameters to be processed.
+        /// </param>
+        private static void EarlyParseHelper(string[] args)
+        {
+            Dbg.Assert(args != null, "Argument 'args' to ParseHelper should never be null");
```
[the remark this answers is not in the document]
Comment (TravisEz13) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: resolved

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 6 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs]: Done #7469.

**Pull request merged**